chore: convert ExtensionsWidget, HeaderPrimary, HeaderSecondary into TS

danny5577 opened the pull request:
<!--
IMPORTANT: We applaud pull requests, they excite us every single time. As we have an obligation to maintain a healthy code standard and quality, we take sufficient time for reviews. Please do create a separate pull request per change/issue/feature; we will ask you to split bundled pull requests.
-->

**Changes proposed in this pull request:**
Converts following files into TS: ExtensionsWidget, HeaderPrimary, HeaderSecondary

**Reviewers should focus on:**
All the feedback is very appreciated, as i am still learning TS

**Necessity**

- [ ] Has the problem that is being solved here been clearly explained?
- [ ] If applicable, have various options for solving this problem been considered?
- [ ] For core PRs, does this need to be in core, or could it be in an extension?
- [ ] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [ ] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).
- [ ] Core developer confirmed locally this works as intended.
- [ ] Tests have been added, or are not appropriate here.

**Required changes:**

- [ ] Related documentation PR: (Remove if irrelevant)
- [ ] Related core extension PRs: (Remove if irrelevant)

Commit:
chore: convert ExtensionsWidget,HeaderPrimary,HeaderSecondary to TS

Commit:
Merge branch 'main' of https://github.com/danny5577/framework

Commit:
chore: convert ExtensionsWidget,HeaderPrimary,HeaderSecondary into TS

SychO9 commented on framework/core/js/src/admin/components/ExtensionsWidget.tsx:
```diff
@@ -1,12 +1,19 @@
-import app from '../../admin/app';
-import DashboardWidget from './DashboardWidget';
+import app from '../app';
+import DashboardWidget, {IDashboardWidgetAttrs} from './DashboardWidget';
 import isExtensionEnabled from '../utils/isExtensionEnabled';
 import getCategorizedExtensions from '../utils/getCategorizedExtensions';
 import Link from '../../common/components/Link';
 import icon from '../../common/helpers/icon';
+import { Extension } from '../AdminApplication';
+import type Mithril from 'mithril';
```
Let's import types as only types 👍🏼 

```suggestion
import DashboardWidget from './DashboardWidget';
import isExtensionEnabled from '../utils/isExtensionEnabled';
import getCategorizedExtensions from '../utils/getCategorizedExtensions';
import Link from '../../common/components/Link';
import icon from '../../common/helpers/icon';
import type { Extension } from '../AdminApplication';
import type { IDashboardWidgetAttrs } from './DashboardWidget';
import type Mithril from 'mithril';
```

SychO9 commented on framework/core/js/src/admin/components/ExtensionsWidget.tsx:
```diff
@@ -26,16 +33,16 @@ export default class ExtensionsWidget extends DashboardWidget {
   }
 
-  extensionCategory(category) {
+  extensionCategory(category:any) {
```
I believe `category` is a string here based on its usage within this method

SychO9 commented on framework/core/js/src/admin/components/ExtensionsWidget.tsx:
```diff
@@ -26,16 +33,16 @@ export default class ExtensionsWidget extends DashboardWidget {
       <div className="ExtensionList-Category">
         <h4 className="ExtensionList-Label">{app.translator.trans(`core.admin.nav.categories.${category}`)}</h4>
-        <ul className="ExtensionList">{this.categorizedExtensions[category].map((extension) => this.extensionWidget(extension))}</ul>
+        <ul className="ExtensionList">{this.categorizedExtensions[category].map((extension:any) => this.extensionWidget(extension))}</ul>
```
Could probably use the `Extension` type that was imported here

SychO9 commented on framework/core/js/src/admin/components/ExtensionsWidget.tsx:
```diff
@@ -26,16 +33,16 @@ export default class ExtensionsWidget extends DashboardWidget {
   }
 
-  extensionWidget(extension) {
+  extensionWidget(extension:any) {
```
same here

SychO9 commented on framework/core/js/src/admin/components/HeaderPrimary.tsx:
```diff
@@ -20,10 +22,9 @@ export default class HeaderPrimary extends Component {
-   * @return {ItemList<import('mithril').Children>}
    */
   items() {
-    return new ItemList();
+    const items = new ItemList<Mithril.Children>();
+    return items;
```
let's add an explicit return type to this method and the other items method changed in this PR

Commit:
enabling tests

Commit:
Merge branch 'main' of https://github.com/danny5577/framework

flarum-bot committed:
Bundled output for commit ea53e9c79b43a74f229bbdd15c925fa463b8ae1a
Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]
